### PR TITLE
drivers: adc: stm32: PM support

### PR DESCRIPTION
**Pull request by fouge**
Low power mode with ADC devices: ADC pin and device clock suspended when Power Manager requires it

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -1296,6 +1297,55 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+		/* Set pins to active state */
+		err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+		if (err < 0) {
+			return err;
+		}
+
+		/* enable clock */
+		err = clock_control_on(data->clock, (clock_control_subsys_t)&config->pclken);
+		if (err != 0) {
+			LOG_ERR("Could not enable (LP)UART clock");
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1320]: Fix comment

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -1296,6 +1297,55 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+			return err;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Stop device clock. Note: fixed clocks are not handled yet. */
+		err = clock_control_off(data->clock, (clock_control_subsys_t)&config->pclken);
+		if (err != 0) {
+			LOG_ERR("Could not enable (LP)UART clock");
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1328]: ditto

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -1296,6 +1297,55 @@ static const struct adc_driver_api api_stm32_driver_api = {
...
+			return err;
+		}
+
+		/* Move pins to sleep state */
+		err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
+		if (err == -ENOENT) {
+			/* Warn but don't block PM suspend */
+			LOG_WRN("(LP)UART pinctrl sleep state not available ");
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:1336]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by fouge** [comment]
closing as it doesn't provide any significant gain

**Pull request closed without merging**